In [1]:
import pandas as pd
import numpy as np

In [2]:
kdd = pd.read_csv("data/kddcup_test_data")
kdd

duration protocol_type   service flag  src_bytes  dst_bytes  land  \
0              0           udp   private   SF        105        146     0   
1              0           udp   private   SF        105        146     0   
2              0           udp   private   SF        105        146     0   
3              0           udp   private   SF        105        146     0   
4              0           udp   private   SF        105        146     0   
5              0           udp   private   SF        105        146     0   
6              0           udp  domain_u   SF         29          0     0   
7              0           udp   private   SF        105        146     0   
8              0           udp   private   SF        105        146     0   
9              0           tcp      http   SF        223        185     0   
10             0           udp   private   SF        105        146     0   
11             0           tcp      http   SF        230        260     0   
12             0           udp   private   SF        105        146     0   
13             0           udp   private   SF        105        146     0   
14             1           tcp      smtp   SF       3170        329     0   
15             0           tcp      http   SF        297      13787     0   
16             0           tcp      http   SF        291       3542     0   
17             0           tcp      http   SF        295        753     0   
18             0           udp   private   SF        105        146     0   
19             0           udp   private   SF        105        146     0   
20             0           tcp      http   SF        268       9235     0   
21             0           udp   private   SF        105        146     0   
22             0           tcp      http   SF        223        185     0   
23             0           tcp      http   SF        227       8841     0   
24             0           tcp      http   SF        222      19564     0   
25             0           tcp  ftp_data   SF        740          0     0   
26             0           udp   private   SF        105        146     0   
27             0           tcp  ftp_data   SF      35195          0     0   
28             0           tcp  ftp_data   SF       8325          0     0   
29             0           udp   private   SF        105        146     0   
...          ...           ...       ...  ...        ...        ...   ...   
310999         0           udp   private   SF        105        105     0   
311000         0           udp   private   SF        105        105     0   
311001         0           udp   private   SF        105        105     0   
311002         0           udp   private   SF        105        105     0   
311003         0           udp   private   SF        105        105     0   
311004         0           udp   private   SF        105        105     0   
311005         0           udp   private   SF        105        105     0   
311006         0           udp   private   SF        105        105     0   
311007         0           udp   private   SF        105        105     0   
311008         0           udp   private   SF        105        105     0   
311009         0           udp   private   SF        105        105     0   
311010         0           udp   private   SF        105        105     0   
311011         0           udp   private   SF        105        105     0   
311012         0           udp   private   SF        105        105     0   
311013         0           udp   private   SF        105        105     0   
311014         0           udp   private   SF        105        105     0   
311015         0           udp   private   SF        105        105     0   
311016         0           udp   private   SF        105        105     0   
311017         0           udp   private   SF        105        105     0   
311018         0           udp   private   SF        105        105     0   
3110

In [4]:
pd.unique(kdd.classification.ravel())

array(['normal.', 'snmpgetattack.', 'named.', 'xlock.', 'smurf.',
       'ipsweep.', 'multihop.', 'xsnoop.', 'sendmail.', 'guess_passwd.',
       'saint.', 'buffer_overflow.', 'portsweep.', 'pod.', 'apache2.',
       'phf.', 'udpstorm.', 'warezmaster.', 'perl.', 'satan.', 'xterm.',
       'mscan.', 'processtable.', 'ps.', 'nmap.', 'rootkit.', 'neptune.',
       'loadmodule.', 'imap.', 'back.', 'httptunnel.', 'worm.',
       'mailbomb.', 'ftp_write.', 'teardrop.', 'land.', 'sqlattack.',
       'snmpguess.'], dtype=object)

In [8]:
normal = kdd[kdd.classification == "normal."]

In [11]:
#dos
back= kdd[kdd.classification == "back."]
land = kdd[kdd.classification == "land."]
neptune = kdd[kdd.classification == "neptune."]
pod = kdd[kdd.classification == "pod."]
smurf = kdd[kdd.classification == "smurf."]
teardrop = kdd[kdd.classification == "teardrop."]

snmpgetattack = kdd[kdd.classification == "snmpgetattack."]
apache2 = kdd[kdd.classification == "apache2."]
udpstorm = kdd[kdd.classification == "udpstorm."]
processtable = kdd[kdd.classification == "processtable."]
mailbomb = kdd[kdd.classification == "mailbomb."]



frames = [back,land,neptune,pod,smurf,teardrop,snmpgetattack,apache2, udpstorm,mailbomb,processtable]
dos = pd.concat(frames)

In [10]:
#u2r
loadmodule = kdd[kdd.classification == "loadmodule."]
buffer_overflow = kdd[kdd.classification == "buffer_overflow."]
perl = kdd[kdd.classification == "perl."]
rootkit = kdd[kdd.classification == "rootkit."]

xterm = kdd[kdd.classification == "xterm."]
ps = kdd[kdd.classification == "ps."]
sqlattack = kdd[kdd.classification == "sqlattack."]




frames = [loadmodule,buffer_overflow,perl,rootkit,xterm, ps,sqlattack]
u2r = pd.concat(frames)

In [12]:

#r2l
ftp_write = kdd[kdd.classification == "ftp_write."]
guess_passwd = kdd[kdd.classification == "guess_passwd."]
imap = kdd[kdd.classification == "imap."]
multihop = kdd[kdd.classification == "multihop."]
phf = kdd[kdd.classification == "phf."]
spy = kdd[kdd.classification == "spy."]
warezclient = kdd[kdd.classification == "warezclient."]
warezmaster = kdd[kdd.classification == "warezmaster."]

named = kdd[kdd.classification == "named."]
xlock = kdd[kdd.classification == "xlock."]
xsnoop = kdd[kdd.classification == "xsnoop."]
sendmail = kdd[kdd.classification == "sendmail."]
httptunnel = kdd[kdd.classification == "httptunnel."]
worm = kdd[kdd.classification == "worm."]

frames = [ftp_write,guess_passwd,imap,multihop,phf,spy,warezclient,warezmaster,named,xlock,xsnoop,sendmail,httptunnel,worm]
r2l = pd.concat(frames)


In [13]:

#probe
ipsweep = kdd[kdd.classification == "ipsweep."]
nmap = kdd[kdd.classification == "nmap."]
portsweep = kdd[kdd.classification == "portsweep."]
satan = kdd[kdd.classification == "satan."]

saint = kdd[kdd.classification == "saint."]
mscan = kdd[kdd.classification == "mscan."]
snmpguess = kdd[kdd.classification == "snmpguess."]


frames = [ipsweep,nmap,portsweep,satan,saint,mscan,snmpguess]
probe = pd.concat(frames)

In [14]:
def makeBaskets(array):
    tempArray = []
    for x in range(0, len(array)):
        tempArray.append([])
        # 0 = "low_duration"
        # 0-48 = "medium_duration"
        # >48 = "high_duration"
        if array[x][0]== 0:
            tempArray[x].append("low_duration")
        elif array[x][0] > 0 and array[x][0]<= 100:
            tempArray[x].append("medium_duration")
        elif array[x][0] > 100:
            tempArray[x].append("high_duration")
            
        # 0-520 = low_src_bytes
        # 520 - 3025 = medium_src_bytes
        # >3025 = high_src_bytes
        if array[x][4] <=520:
            tempArray[x].append("low_src_bytes")
        elif array[x][4] > 520 and array[x][4] <= 5000:
            tempArray[x].append("medium_src_bytes")
        elif array[x][4] > 5000:
            tempArray[x].append("high_src_bytes")
            
        # 0 = low_dst_bytes
        # 0-869 = medium_dst_bytes
        # > 869 = high_dst_bytes
        if array[x][5] == 0:
            tempArray[x].append("low_dst_bytes")
        elif array[x][5]> 0 & array[x][5]<= 1000:
            tempArray[x].append("medium_dst_bytes")
        elif array[x][5]>1000:
            tempArray[x].append("high_dst_bytes")
            
        # yes_land or no_land
        if array[x][6] == 0:
            tempArray[x].append("no_land")
        elif array[x][6] == 1:
            tempArray[x].append("yes_land")
            
        # 0 = low_wrong_fragment
        # 0-1 = medium_wrong_fragment
        # >1 = high_wrong_fragment
        if array[x][7] > 0:
            tempArray[x].append("wrong_fragment")
            
        # 0 = low_urgent
        #>0 = high_urgent
        
        if array[x][8] > 0:
            tempArray[x].append("urgent")
            
        # 0 = low_hot
        # 0-1 = medium_hot
        # >0 = high_hot
        if array[x][9] >0:
            tempArray[x].append("hot")
            
        # 0 = low_num_failed_logins
        # 0-1 = medium_num_failed_logins
        # >1 = high_num_failed_logins
        if array[x][10] > 0:
            tempArray[x].append("failed_logins")
    
            
        #no_logged_in
        #yes_logged_in
        if array[x][11] == 0:
            tempArray[x].append("no_logged_in")
        elif array[x][11] == 1:
            tempArray[x].append("yes_logged_in")
            
        # 0 = low_num_compromised
        # >0 = high_num_compromised
        if array[x][12] > 0:
            tempArray[x].append("num_compromised")

        # 0 = low_root_shell
        # >0 = high_root_shell
        if array[x][13] >0:
            tempArray[x].append("root_shell")
            
        #0 = low_su_attempted
        #>1 = high_su_attempted
        if array[x][14] >0:
            tempArray[x].append("su_attempted")
            
        #0 = low_num_root
        # >0 = high_num_root
        if array[x][15] >0:
            tempArray[x].append("high_num_root")
            
        # 0 = low_file_creations
        #>1 = high_file_creations
        if array[x][16] >0:
            tempArray[x].append("file_creations")
            
        # 0 = low_num_shells
        # >1 = high_num_shells
        if array[x][17] >0:
            tempArray[x].append("high_num_shells")
            
        # 0 = low_num_access_files
        #>1 = high_num_access_files
        if array[x][18] >0:
            tempArray[x].append("high_num_access_files")
            
        # 0 = low_num_outbound_cmds
        #>1 = high_num_outbound_cmds
        if array[x][19] >0:
            tempArray[x].append("high_num_outbound_cmds")
             
        #'is_hot_login',
        if array[x][20] == 0:
            tempArray[x].append("no_hot_login")
        elif array[x][20] ==1:
            tempArray[x].append("yes_host_login")
 
        #'is_guest_login',
        if array[x][21] == 0:
            tempArray[x].append("no_guest_login")
        elif array[x][21] ==1:
            tempArray[x].append("yes_guest_login")
            
        #0-332 low_count
        #332-510 medium_count
        #510-511 high_count
        if array[x][22] < 332:
            tempArray[x].append("low_count")
        elif array[x][22]>= 332:
            tempArray[x].append("high_count")
            
        #0-293 = low_srv_count
        #293-510 = medium_srv_count
        #>=510 = high_srv_count
        if array[x][23] < 292:
            tempArray[x].append("low_srv_count")
        elif array[x][23]>= 292:
            tempArray[x].append("high_srv_count")

            
        # 0 = low_serror_rate
        # 0 - 0.17 = medium_serror_rate
        # > 0.17 = high_serror_rate
        
        if array[x][24] >0:
            tempArray[x].append("high_serror_rate")
            
        # 0 = low_srv_serror_rate
        # 0 - 0.17 = medium_srv_serror_rate
        # > 0.17 = high_srv_serror_rate
        if array[x][25] >0:
            tempArray[x].append("high_srv_serror_rate")
            
        # 0 = low rerror_rate
        # 0 - 0.05 = medium_rerror_rate
        #>.05 = high_rerror_rate
        if array[x][26] >0:
            tempArray[x].append("high_rerror_rate")
            
        # 0 = low_srv_rerror_rate
        # 0 - 0.05 = medium_srv_rerror_rate
        #>.05 = high_srv_rerror_rate
        if array[x][27] >0:
            tempArray[x].append("high_srv_rerror_rate")

        # 0- .79 = low_same_srv_rate
        #.79-1 = medium_same_srv_rate
        # 1 = high_same_srv_rate
        if array[x][28] <1:
            tempArray[x].append("low_same_srv_rate")
            
        # 0 = low_diff_srv_rate
        # 0-.02 = medium_diff_srv_rate
        # > .02 = high_diff_srv_rate
        if array[x][29] >0:
            tempArray[x].append("high_diff_srv_rate")

            
        # 0 = low_srv_diff_host_rate
        # 0-.02 = medium_srv_diff_host_rate
        # > .02 = high_srv_diff_host_rate
        if array[x][30] >0:
            tempArray[x].append("high_srv_diff_host_rate")
    
        #255 = high_dst_host_count
        #255-232 = medium_dst_host_count
        # 0-232 = low_dst_host_count
        if array[x][31] <232:
            tempArray[x].append("low_dst_host_count")
        elif array[x][31] >=232: 
            tempArray[x].append("high_dst_host_count")
    
        #255 = high_dst_host_srv_count
        #255-188 = medium_dst_host_srv_count
        # 0-188 = low_dst_host_srv_count
        if array[x][32] <188:
            tempArray[x].append("low_dst_host_srv_count")
        elif array[x][32] >=188:
            tempArray[x].append("high_dst_host_srv_count")
        
        
        # 0-.75 low_dst_host_same_srv_rate
        # .75 -1 high_dst_host_sam_srv_rate
        
        if array[x][33]<.75:
            tempArray[x].append("low_dst_host_same_srv_rate")
        elif array[x][33] >= .75:
            tempArray[x].append("high_dst_host_sam_srv_rate")
            
        # 0 low_dst_host_diff_srv_rate
        # >0 high_dst_host_diff_srv_rate
        if array[x][34]>0:
            tempArray[x].append("high_dst_host_diff_srv_rate")
    
        #<.6 low_dst_host_same_src_port_rate
        #>=.6 high_dst_host_same_src_port_rate
        if array[x][35]<.6:
            tempArray[x].append("low_dst_host_same_src_port_rate")
        elif array[x][35] >=.6:
            tempArray[x].append("high_dst_host_same_src_port_rate")
        
        # <.01 low_dst_host_srv_diff_host_rate
        # >.01 high_dst_host_srv_diff_host_rate
        if array[x][36]>0:
            tempArray[x].append("high_dst_host_srv_diff_host_rate")

            
        # <.17 low_dst_host_serror_rate
        # >.17 high_dst_host_serror_rate
        if array[x][37]<.17:
            tempArray[x].append("low_dst_host_serror_rate")
        elif array[x][37] >=.17:
            tempArray[x].append("high_dst_host_serror_rate")
            
        # <.17 low_dst_host_srv_serror_rate
        #>.17 high_dst_host_srv_serror_rate
        if array[x][38]<.17:
            tempArray[x].append("low_dst_host_srv_serror_rate")
        elif array[x][38] >=.17:
            tempArray[x].append("high_dst_host_srv_serror_rate")
        
        # <.05 low_dst_host_rerror_rate
        # >.05 high_dst_host_rerror_rate
        if array[x][39]>0:
            tempArray[x].append("low_dst_host_srv_serror_rate")
            
        #<.05 low_dst_host_srv_rerror_rate
        #>.05 high_dst_host_srv_rerror_rate
        if array[x][40]>0:
            tempArray[x].append("low_dst_host_srv_serror_rate")
    return tempArray

In [15]:
def load_dataset():
    "Load the sample dataset."
    return [[1, 3, 4], [2, 3, 5], [1, 2, 3, 5], [2, 5]]


def createC1(dataset):
    "Create a list of candidate item sets of size one."
    c1 = []
    for transaction in dataset:
        for item in transaction:
            if not [item] in c1:
                c1.append([item])
    c1.sort()
    #frozenset because it will be a ket of a dictionary.
    return map(frozenset, c1)


def scanD(dataset, candidates, min_support):
    "Returns all candidates that meets a minimum support level"
    sscnt = {}
    for tid in dataset:
        for can in candidates:
            if can.issubset(tid):
                sscnt.setdefault(can, 0)
                sscnt[can] += 1

    num_items = float(len(dataset))
    retlist = []
    support_data = {}
    for key in sscnt:
        support = sscnt[key] / num_items
        if support >= min_support:
            retlist.insert(0, key)
        support_data[key] = support
    return retlist, support_data


def aprioriGen(freq_sets, k):
    "Generate the joint transactions from candidate sets"
    retList = []
    lenLk = len(freq_sets)
    for i in range(lenLk):
        for j in range(i + 1, lenLk):
            L1 = list(freq_sets[i])[:k - 2]
            L2 = list(freq_sets[j])[:k - 2]
            L1.sort()
            L2.sort()
            if L1 == L2:
                retList.append(freq_sets[i] | freq_sets[j])
    return retList


def apriori(dataset, minsupport=0.5):
    "Generate a list of candidate item sets"
    C1 = createC1(dataset)
    D = map(set, dataset)
    L1, support_data = scanD(D, C1, minsupport)
    L = [L1]
    k = 2
    while (len(L[k - 2]) > 0):
        Ck = aprioriGen(L[k - 2], k)
        Lk, supK = scanD(D, Ck, minsupport)
        support_data.update(supK)
        L.append(Lk)
        k += 1

    return L, support_data

In [16]:
u2r_group=[frozenset({'low_dst_host_same_src_port_rate'}),
 frozenset({'low_src_bytes'}),
 frozenset({'yes_logged_in'}),
 frozenset({'low_dst_host_count'}),
 frozenset({'low_count'}),
 frozenset({'medium_dst_bytes'}),
 frozenset({'root_shell'}),
 frozenset({'low_srv_count'}),
 frozenset({'low_dst_host_srv_count'}),
 frozenset({'hot'}),
 frozenset({'high_dst_host_sam_srv_rate'})]

In [17]:
normal_group=[frozenset({'low_dst_host_same_src_port_rate'}),
 frozenset({'low_duration'}),
 frozenset({'high_dst_host_srv_diff_host_rate'}),
 frozenset({'low_src_bytes'}),
 frozenset({'yes_logged_in'}),
 frozenset({'low_dst_host_count'}),
 frozenset({'low_count'}),
 frozenset({'medium_dst_bytes'}),
 frozenset({'high_dst_host_srv_count'}),
 frozenset({'low_srv_count'}),
 frozenset({'high_dst_host_sam_srv_rate'})]

In [18]:
dos_group = [frozenset({'low_duration'}),
 frozenset({'no_logged_in'}),
 frozenset({'medium_src_bytes'}),
 frozenset({'high_dst_host_same_src_port_rate'}),
 frozenset({'high_dst_host_srv_count'}),
 frozenset({'low_dst_bytes'}),
 frozenset({'high_dst_host_count'}),
 frozenset({'high_srv_count'}),
 frozenset({'high_count'}),
 frozenset({'high_dst_host_sam_srv_rate'})]

In [19]:
r2l_group = [frozenset({'high_dst_host_srv_diff_host_rate'}),
 frozenset({'low_src_bytes'}),
 frozenset({'yes_logged_in'}),
 frozenset({'low_duration'}),
 frozenset({'low_dst_host_count'}),
 frozenset({'low_count'}),
 frozenset({'high_dst_host_same_src_port_rate'}),
 frozenset({'low_dst_bytes'}),
 frozenset({'low_srv_count'}),
 frozenset({'low_dst_host_srv_count'}),
 frozenset({'high_dst_host_sam_srv_rate'})]

In [20]:
probe_group =[frozenset({'low_duration'}),
 frozenset({'low_src_bytes'}),
 frozenset({'no_logged_in'}),
 frozenset({'high_rerror_rate'}),
 frozenset({'high_dst_host_diff_srv_rate'}),
 frozenset({'low_count'}),
 frozenset({'low_dst_host_same_srv_rate'}),
 frozenset({'high_dst_host_same_src_port_rate'}),
 frozenset({'low_dst_bytes'}),
 frozenset({'high_dst_host_count'}),
 frozenset({'low_srv_count'}),
 frozenset({'low_dst_host_srv_count'}),
 frozenset({'high_srv_rerror_rate'})]

In [21]:
def make_guess(test_signature):
    C_test = []
    C_test.append(test_signature)
    C_test = createC1(C_test)
    dif1 = len(set(C_test) & set(normal_group))
    dif2 = len(set(C_test) & set(u2r_group))
    dif3 = len(set(C_test) & set(dos_group))
    dif4 = len(set(C_test) & set(r2l_group))
    dif5 = len(set(C_test) & set(probe_group))
    
    max = dif1
    result = "normal"
    if dif2 > max:
        max = dif2
        result = "u2r"
    if dif3 > max:
        max = dif3
        result = "dos"
    if dif4 > max:
        max = dif4
        result = "r2l"
    if dif5 > max:
        result = "probe"
    return result
    

In [22]:
normal_basket_data = normal.values
dos_basket_data = dos.values
u2r_basket_data = u2r.values
r2l_basket_data = r2l.values
probe_basket_data = probe.values

In [23]:
normal_baskets = makeBaskets(normal_basket_data)
u2r_baskets = makeBaskets(u2r_basket_data)
dos_baskets = makeBaskets(dos_basket_data)
r2l_baskets = makeBaskets(r2l_basket_data)
probe_baskets = makeBaskets(probe_basket_data)

In [90]:
def run_test(basket):
    
    normal = 0
    u2r = 0
    dos = 0
    r2l = 0
    probe = 0
    for x in range(len(basket)):
        result = make_guess(basket[x])
        if result == "normal":
            normal += 1
        elif result == "u2r":
            u2r += 1
        elif result == "dos":
            dos += 1
        elif result == "r2l":
            r2l +=1
        else:
            probe += 1
    return normal,probe,dos,u2r,r2l

In [115]:
normal,probe,dos,u2r,r2l = run_test(normal_baskets)

In [116]:
normal_results = [normal,probe,dos,u2r,r2l]

In [117]:
normal,probe,dos,u2r,r2l= run_test(u2r_baskets)

In [118]:
u2r_results = [normal,probe,dos,u2r,r2l]

In [119]:
normal,probe,dos,u2r,r2l = run_test(dos_baskets)

In [120]:
dos_results = [normal,probe,dos,u2r,r2l]

In [121]:
normal,probe,dos,u2r,r2l = run_test(r2l_baskets)

In [122]:
r2l_results = [normal,probe,dos,u2r,r2l]

In [123]:
normal,probe,dos,u2r,r2l = run_test(probe_baskets)

In [124]:
probe_results = [normal,probe,dos,u2r,r2l]

In [125]:
total_results = np.array([normal_results,probe_results,dos_results,u2r_results,r2l_results])

In [126]:
total_results

array([[ 52681,   6089,    660,    593,    570],
       [  1495,   4667,      4,      2,    404],
       [ 14330,  59026, 163968,    171,     99],
       [     1,     11,      0,     52,      6],
       [  3809,    622,      0,    803,    966]])

<!DOCTYPE html>
<html>
<head>
<title>HTML Tables</title>
</head>
<body>
<p>Association Confusion Matrix </p>
<table border="1">
<tr>
<td><p>Predicted</p><p> Actual</p></td>
<td>Normal</td>
<td>Probe</td>
<td>DOS</td>
<td>U2R</td>
<td>R2L</td>
<td>Correct %</td>
</tr>
<tr>
<td>normal</td>
<td>52681</td>
<td>6089</td>
<td>660</td>
<td>593</td>
<td>570</td>
<td>86.9%</td>
</tr>
<tr>
<td>probe</td>
<td>1495</td>
<td>4667</td>
<td>4</td>
<td>2</td>
<td>404</td>
<td>71.0%</td>

</tr>
<tr>
<td>DOS</td>
<td>14330</td>
<td>59026</td>
<td>163968</td>
<td>171</td>
<td>99</td>
<td>69.0%</td>
</tr>
<tr>
<td>U2r</td>
<td>1</td>
<td>11</td>
<td>0</td>
<td>52</td>
<td>6</td>
<td>74.2%</td>
</tr>
<tr>
<td>R2L</td>
<td>3809</td>
<td>622</td>
<td>0</td>
<td>803</td>
<td>966</td>
<td>15.5%</td>
</tr>
<tr>
<td>Correct %</td>
<td>72.8%</td>
<td>6.6%</td>
<td>99.5%</td>
<td>3.2%</td>
<td>47.2%</td>
<td></td>
</tr>
</table>
</body>
</html>